File: app/src/main/java/com/example/jacob/myapplication/activities/MainActivity.java
package com.example.jacob.myapplication.activities;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.os.ParcelFileDescriptor;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import com.example.jacob.myapplication.Constants;
import com.example.jacob.myapplication.tasks.CreateDataTask;
import com.example.jacob.myapplication.db.DataDBHandler;
import com.example.jacob.myapplication.db.SavedDataAdapter;
import com.example.jacob.myapplication.logic.IConversationData;
import com.example.jacob.myapplication.R;
import com.nononsenseapps.filepicker.Utils;

import java.io.FileNotFoundException;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        Constants.dbHandler = new DataDBHandler(this);

        SavedDataAdapter sva = new SavedDataAdapter(this, Constants.dbHandler.getAllData());
        ListView l = findViewById(R.id.saved_analysis_list);
        l.setAdapter(sva);

        l.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Constants.conversationData =(IConversationData) parent.getItemAtPosition(position);
                Intent intent = new Intent(view.getContext(), ResultsActivity.class);
                startActivity(intent);
            }
        });

        if (getIntent().getAction().equals(Intent.ACTION_VIEW)){
            Uri returnUri = getIntent().getData();
            ParcelFileDescriptor mInputPFD = null;
            try {
                mInputPFD = getContentResolver().openFileDescriptor(returnUri, "r");
            } catch (FileNotFoundException e) {
                e.printStackTrace();
                Log.e("MainActivity", "File not found.");
            }
            new CreateDataTask(this, returnUri.getLastPathSegment()).execute(mInputPFD);
        }

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.

        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public void buttonClick(View view){

        Context context = getApplicationContext();
        // This always works
        Intent i = new Intent(context, MyPickerActivity.class);
        // This works if you defined the intent filter
        // Intent i = new Intent(Intent.ACTION_GET_CONTENT);

        // Set these depending on your use case. These are the defaults.
        i.putExtra(MyPickerActivity.EXTRA_ALLOW_MULTIPLE, false);
        i.putExtra(MyPickerActivity.EXTRA_ALLOW_CREATE_DIR, false);
        i.putExtra(MyPickerActivity.EXTRA_MODE, MyPickerActivity.MODE_FILE);
        i.putExtra(MyPickerActivity.EXTRA_SINGLE_CLICK, false);
        i.putExtra(MyPickerActivity.EXTRA_ALLOW_EXISTING_FILE, true);

        // Configure initial directory by specifying a String.
        // You could specify a String like "/storage/emulated/0/", but that can
        // dangerous. Always use Android's API calls to get paths to the SD-card or
        // internal memory.
        i.putExtra(MyPickerActivity.EXTRA_START_PATH, Environment.getExternalStorageDirectory().getPath());
        startActivityForResult(i, 1);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {

        if (resultCode == Activity.RESULT_OK) {
            List<Uri> files = Utils.getSelectedFilesFromResult(data);
            Uri uri = files.get(0);
            ParcelFileDescriptor mInputPFD = null;
            try {
                mInputPFD = getContentResolver().openFileDescriptor(uri, "r");
            } catch (FileNotFoundException e) {
                e.printStackTrace();
                Log.e("MainActivity", "File not found.");
            }
            new CreateDataTask(this, uri.getLastPathSegment()).execute(mInputPFD);
        }
    }
}

File: app/src/main/java/com/example/jacob/myapplication/activities/MainActivity.kt
package com.example.jacob.myapplication.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Environment
import android.os.ParcelFileDescriptor
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import com.example.jacob.myapplication.Constants
import com.example.jacob.myapplication.R
import com.example.jacob.myapplication.db.DataDBHandler
import com.example.jacob.myapplication.db.SavedDataAdapter
import com.example.jacob.myapplication.logic.IConversationData
import com.example.jacob.myapplication.tasks.CreateDataTask
import com.nononsenseapps.filepicker.FilePickerActivity
import com.nononsenseapps.filepicker.Utils
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import java.io.FileNotFoundException

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        Constants.dbHandler = DataDBHandler(this)

        val sva = SavedDataAdapter(this, Constants.dbHandler.allData)
        saved_analysis_list.adapter = sva

        saved_analysis_list.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, _ ->
            Constants.conversationData = parent.getItemAtPosition(position) as IConversationData
            val intent = Intent(view.context, ResultsActivity::class.java)
            startActivity(intent)
        }

        if (intent.action == Intent.ACTION_VIEW) {
            val returnUri = intent.data
            var mInputPFD: ParcelFileDescriptor? = null
            try {
                mInputPFD = contentResolver.openFileDescriptor(returnUri!!, "r")
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
                Log.e("MainActivity", "File not found.")
            }

            CreateDataTask(this, returnUri!!.lastPathSegment).execute(mInputPFD)
        }

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId

        return if (id == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)

    }

    fun buttonClick(view: View) {

        val context = applicationContext
        // This always works
        val i = Intent(context, MyPickerActivity::class.java)
        // This works if you defined the intent filter
        // Intent i = new Intent(Intent.ACTION_GET_CONTENT);

        // Set these depending on your use case. These are the defaults.
        i.putExtra(FilePickerActivity.EXTRA_ALLOW_MULTIPLE, false)
        i.putExtra(FilePickerActivity.EXTRA_ALLOW_CREATE_DIR, false)
        i.putExtra(FilePickerActivity.EXTRA_MODE, FilePickerActivity.MODE_FILE)
        i.putExtra(FilePickerActivity.EXTRA_SINGLE_CLICK, false)
        i.putExtra(FilePickerActivity.EXTRA_ALLOW_EXISTING_FILE, true)

        // Configure initial directory by specifying a String.
        // You could specify a String like "/storage/emulated/0/", but that can
        // dangerous. Always use Android's API calls to get paths to the SD-card or
        // internal memory.
        i.putExtra(FilePickerActivity.EXTRA_START_PATH, Environment.getExternalStorageDirectory().path)
        startActivityForResult(i, 1)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {

        if (resultCode == Activity.RESULT_OK) {
            val files = Utils.getSelectedFilesFromResult(data)
            val uri = files[0]
            var mInputPFD: ParcelFileDescriptor? = null
            try {
                mInputPFD = contentResolver.openFileDescriptor(uri, "r")
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
                Log.e("MainActivity", "File not found.")
            }

            CreateDataTask(this, uri.lastPathSegment).execute(mInputPFD)
        }
    }
}
